libzfs/src/vdev.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) 2018 Intel Corporation. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+extern crate libzfs_sys as sys;
+
+use nvpair;
+use std::ffi::CStr;
+use libzfs_error::{LibZfsError, Result};
+use std::io::{Error, ErrorKind};
+
+/// Represents vdevs
+/// The enum starts at Root and is recursive.
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub enum VDev {
+    Mirror {
+        children: Vec<VDev>,
+        is_log: Option<bool>,
+    },
+    RaidZ {
+        children: Vec<VDev>,
+    },
+    Replacing {
+        children: Vec<VDev>,
+    },
+    Root {
+        children: Vec<VDev>,
+        spares: Vec<VDev>,
+        cache: Vec<VDev>,
+    },
+    Disk {
+        guid: Option<String>,
+        state: String,
+        path: String,
+        dev_id: Option<String>,
+        phys_path: Option<String>,
+        whole_disk: Option<bool>,
+        is_log: Option<bool>,
+    },
+    File {
+        guid: Option<String>,
+        state: String,
+        path: String,
+        is_log: Option<bool>,
+    },
+}
+
+pub fn enumerate_vdev_tree(tree: &nvpair::NvList) -> Result<VDev> {
+    let tmp = tree.lookup_string(sys::zpool_config_type())?;
+    let x = tmp.as_bytes_with_nul();
+
+    fn get_children(tree: &nvpair::NvList) -> Result<Vec<VDev>> {
+        tree.lookup_nv_list_array(sys::zpool_config_children())?
+            .iter()
+            .map(enumerate_vdev_tree)
+            .collect()
+    }
+
+    fn get_spares(tree: &nvpair::NvList) -> Result<Vec<VDev>> {
+        let spares = tree.lookup_nv_list_array(sys::zpool_config_spares());
+
+        match spares {
+            Ok(x) => x.iter().map(enumerate_vdev_tree).collect(),
+            Err(_) => Ok(vec![]),
+        }
+    }
+
+    fn get_cache(tree: &nvpair::NvList) -> Result<Vec<VDev>> {
+        let cache = tree.lookup_nv_list_array(sys::zpool_config_l2cache());
+
+        match cache {
+            Ok(x) => x.iter().map(enumerate_vdev_tree).collect(),
+            Err(_) => Ok(vec![]),
+        }
+    }
+
+    fn lookup_tree_str(tree: &nvpair::NvList, name: String) -> Result<Option<String>> {
+        let x = tree.lookup_string(name);
+
+        match x {
+            Ok(x) => Ok(Some(x.into_string()?)),
+            Err(_) => Ok(None),
+        }
+    }
+
+    fn lookup_is_log(tree: &nvpair::NvList) -> Option<bool> {
+        tree.lookup_uint64(sys::zpool_config_is_log())
+            .map(|x| x == 1)
+            .ok()
+    }
+
+    fn lookup_guid(tree: &nvpair::NvList) -> Option<String> {
+        tree.lookup_uint64(sys::zpool_config_guid())
+            .map(|x| format!("{:#018X}", x))
```

Comment:
What is #018X?

Reply:
It represents padding width + hex formatting. Here is the patch that it was changed in: https://github.com/intel-hpdd/rust-libzfs/pull/33